crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -216,74 +216,139 @@ impl<'db> InnerIntersectionBuilder<'db> {
     }
 
     /// Adds a positive type to this intersection.
-    fn add_positive(&mut self, db: &'db dyn Db, ty: Type<'db>) {
+    fn add_positive(&mut self, db: &'db dyn Db, new_positive: Type<'db>) {
         // TODO `Any`/`Unknown`/`Todo` actually should not self-cancel
-        match ty {
-            Type::Intersection(inter) => {
-                let pos = inter.positive(db);
-                let neg = inter.negative(db);
-                self.positive.extend(pos.difference(&self.negative));
```

Comment:
I did that change locally. It works fine, but I'm not sure I like it. 

> It might require one additional check to avoid adding exactly the same type to the positive or negative side that is already there, but that's a one-line check, we don't need all the `IndexSet` overhead for that.

But that's exactly what a set is designed for, right? We don't neet `OrderSet` — we could also use `HashSet` (because I assume we do not need to maintain insertion order for intersections, just for unions?). But turning every insertion/pop into an O(n) operation feels wrong. And we would probably want to wrap the `Vec` in a new datastructure for that, to make sure noone calls `.push` directly.

Reply:
I think it is already "wrapped" in the `IntersectionType` / `IntersectionBuilder`, so I'm not sure we'd need to wrap it again; the intersection builder code is already quite subtle, and a direct push to an internal vector that doesn't go through `add_negative` / `add_positive` would I think stand out as a bug more than many other bugs we might introduce.

> turning every insertion/pop into an O(n) operation feels wrong

For insertions, they already are; I assume we could put this check inside our existing loop over `self.positive` (in `add_positive`) and over `self.negative` (in `add_negative`). But you're right this might make the removals we have to do in the builder sometimes a bit more expensive? Though I expect that for small intersections, hashing overhead might cost more.

I'm really just dumping some thoughts here for future. If you feel the change doesn't work out well, I'm happy to leave it as-is for now. This tradeoff between the hashing/allocation overhead of a hashset and the extra cost of pops is something that should really be decided by empirical performance data at some point when we do more focused optimization of intersection builder, not by guesses.

> because I assume we do not need to maintain insertion order for intersections, just for unions

Good point! At least that is true until/unless we have a PEP for intersection types that gives users a way to spell them directly in annotations; then we might want to preserve their spelling when we can.

Reply:
> I think it is already "wrapped" in the `IntersectionType` / `IntersectionBuilder`, so I'm not sure we'd need to wrap it again

There are several direct calls to `OrderSet::insert` in **`Inner`**`IntersectionBuilder`, so if we replace that with `Vec`, we would need to wrap it again, or make sure we only call a helper function (`self.push_positive(type)`) instead of calling `self.positive.push(…)` directly.

> > turning every insertion/pop into an O(n) operation feels wrong
> 
> For insertions, they already are

(Oops. I meant to say "every insertion/push" there, not "pop")

Why? [`OrderSet::insert`](https://docs.rs/ordermap/latest/ordermap/set/struct.OrderSet.html#method.insert) is O(1), amortized.

> This tradeoff between the hashing/allocation overhead of a hashset and the extra cost of pops is something that should really be decided by empirical performance data

Yes, please! :smile: I think O(n) analysis might be completely misguided here. Even if someone has 10 nested `if` statement and manages to construct an intersection type that really has 10 elements inside (Maybe `int & ~Literal[1] & ~Literal[2] & … & ~Literal[9]`), an O(n²) operation wouldn't be the end of the world. My feeling is that we don't burn any bridges by not doing that directly, but I'd be glad to write some benchmarks and include them in this PR.

Reply:
I wouldn't spend time on benchmarks for this right now. Let's leave it as is, and revisit when our optimization work can be directed and motivated by profiles from real-world large codebases.

Reply:
ok